### PR TITLE
provider/aws: Add CORS settings to S3 bucket

### PR DESCRIPTION
**Pull request by kjmkznr**
This commit is allows to [CORS(Cross-Origin Resource Sharing)](https://docs.aws.amazon.com/AmazonS3/latest/dev/cors.html) rule configuration to S3 bucket resource.

```
resource "aws_s3_bucket" "b" {
    ....
    cors_rule {
        allowed_headers = ["*"]
        allowed_methods = ["PUT","POST"]
        allowed_origins = ["https://s3-website-test.hashicorp.com"]
        expose_headers = ["ETag"]
        max_age_seconds = 3000
    }
    ....
}
```

### CODE REVIEW

**Comment by bizenn** [comment]
:+1:

**Comment by 2matz** [comment]
:+1:

**Comment by f440** [comment]
:+1:

**Comment by yujideveloper** [comment]
:+1:

**Comment by mitto** [comment]
:+1:

**Comment by corrupt952** [comment]
:+1:

File: builtin/providers/aws/resource_aws_s3_bucket.go [hunk 1]
```diff
@@ -221,6 +260,25 @@ func resourceAwsS3BucketRead(d *schema.ResourceData, meta interface{}) error {
...
+		for _, ruleObject := range cors.CORSRules {
+			rule := make(map[string]interface{})
+			rule["allowed_headers"] = ruleObject.AllowedHeaders
+			rule["allowed_methods"] = ruleObject.AllowedMethods
+			rule["allowed_origins"] = ruleObject.AllowedOrigins
+			rule["expose_headers"] = ruleObject.ExposeHeaders
+			rule["max_age_seconds"] = ruleObject.MaxAgeSeconds
+			rules = append(rules, rule)
+		}
+		d.Set("cors_rule", rules)
```
Comment (catsby) [inline builtin/providers/aws/resource_aws_s3_bucket.go]: Can we check for errors here? We've found that setting non-primitives, there can be an error and it's best we inspect/log that

```go
if err := d.Set("cors_rule", rules); err != nil {
  //log
}
```

File: builtin/providers/aws/resource_aws_s3_bucket.go [hunk 2]
```diff
@@ -168,6 +201,12 @@ func resourceAwsS3BucketUpdate(d *schema.ResourceData, meta interface{}) error {
 		}
 	}
 
+	if d.HasChange("cors_rule") {
+		if err := resourceAwsS3BucketCorsUpdate(s3conn, d); err != nil {
+			return err
+		}
+	}
+
 	if d.HasChange("website") {
```
Comment (catsby) [inline builtin/providers/aws/resource_aws_s3_bucket.go]: should we remove the CORs rules if the `website` is removed?

**Comment by catsby** [comment]
Hey @kjmkznr – so sorry for the delay! I do have a question regarding removing the CORs rules if we remove the website, but other than that this looks great! 

Let me know what you think

**Comment by kjmkznr** [comment]
@catsby 
Thanks reviews!

I think, do not need remove CORS rules when website removed.
CORS and website API is independent each other.
We can set CORS rules without website configurations.

File: builtin/providers/aws/resource_aws_s3_bucket.go (hunk 1 above)
Comment (kjmkznr) [inline builtin/providers/aws/resource_aws_s3_bucket.go]: Added.

**Comment by catsby** [comment]
Thanks!

**Comment by kjmkznr** [comment]
@catsby 

Thanks!!

**Pull request merged**